### PR TITLE
8342090: Infer::IncorporationBinaryOp::equals can produce side-effects

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
Type inference uses a cache to store incorporation operations already done. This way we can avoid redoing operations that once done won't produce any change. This can reduce the compilation time when the number of inference variables involved is not trivial. The elements in the cache are implemented with class `com.sun.tools.javac.comp.Infer::IncorporationBinaryOp` the problem here is that the equals method of this class can produce side effects, implying that after a comparison the compared objects can mutate. This patch is fixing this issue,

TIA

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8288590: javac failure: incompatible types: cannot infer type arguments due to Object.hashCode collision`

### Issues
 * [JDK-8342090](https://bugs.openjdk.org/browse/JDK-8342090): Infer::IncorporationBinaryOp::equals can produce side-effects (**Bug** - P4)
 * [JDK-8288590](https://bugs.openjdk.org/browse/JDK-8288590): javac failure: incompatible types: cannot infer type arguments due to Object.hashCode collision (**Bug** - P3)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Contributors
 * Maurizio Cimadamore `<mcimadamore@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21651/head:pull/21651` \
`$ git checkout pull/21651`

Update a local copy of the PR: \
`$ git checkout pull/21651` \
`$ git pull https://git.openjdk.org/jdk.git pull/21651/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21651`

View PR using the GUI difftool: \
`$ git pr show -t 21651`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21651.diff">https://git.openjdk.org/jdk/pull/21651.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21651#issuecomment-2430492920)

### CODE REVIEW

**Comment by vicente-romero-oracle** [comment]
still waiting for the mach5 to finish

**Review by mcimadamore — COMMENTED** [review]
Looks good. I've suggested a couple of simplifications.
I believe we need another way to test. I think, for now, let's include the tests we have in related JBS issues, and run them with both normal and alternate (`-XX:hash=2`) strategies. Then, in a separate JBS issue, maybe figure out how to run a subset of tests automatically, using the alternate hashing.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java [hunk 1]
```diff
@@ -3328,6 +3328,10 @@ public Type subst(Type t, List<Type> from, List<Type> to) {
         return t.map(new Subst(from, to));
     }
 
+    /* this class won't substitute all types for example UndetVars are never substituted, this is like this
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:3331]:
```suggestion
    /* this class won't substitute all types for example UndetVars are never substituted, this is
```

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java [hunk 2]
```diff
@@ -3328,6 +3328,10 @@ public Type subst(Type t, List<Type> from, List<Type> to) {
...
+     * by design as UndetVars are used locally during inference and shouldn't scape from inference rutines,
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:3332]:
```suggestion
     * by design as UndetVars are used locally during inference and shouldn't escape from inference routines,
```

**Review by mcimadamore — APPROVED** [review]
Looks very good - thanks!

**Review by mcimadamore — APPROVED** [review]

**Comment by gilles-duboscq** [comment]
Thank you for looking into this!

**Pull request closed without merging**